daemons/ipa-kdb/ipa_kdb_principals.c

```diff
@@ -1295,163 +1298,203 @@ krb5_error_code ipadb_get_principal(krb5_context kcontext,
+                                                 realm->length,
+                                                 &trusted_realm);
     }
 
-    kerr = ipadb_parse_ldap_entry(kcontext, principal, lentry, entry, &pol);
     if (kerr != 0) {
         goto done;
     }
 
-    if (pol) {
-        kerr = ipadb_fetch_tktpolicy(kcontext, lentry, *entry, pol);
+    /* This is a known trusted realm. Issue a referral depending on whether this
+     * is client or server referral request */
+    if (flags & KRB5_KDB_FLAG_CLIENT_REFERRALS_ONLY) {
+        /* client referral out of realm, set next realm. */
+        kerr = krb5_set_principal_realm(kcontext, norm_princ, trusted_realm);
         if (kerr != 0) {
             goto done;
         }
+        kentry = calloc(1, sizeof(krb5_db_entry));
+        if (!kentry) {
+            kerr = ENOMEM;
+            goto done;
+        }
+
+        kentry->princ = norm_princ;
+        norm_princ = NULL;
+        *entry = kentry;
+
+        goto done;
+    }
+
+    if (flags & KRB5_KDB_FLAG_INCLUDE_PAC) {
+        /* TGS request where KDC wants to generate PAC
+         * but the principal is out of our realm */
+        kerr = KRB5_KDB_NOENTRY;
+        goto done;
+    }
+
+    /* server referrals: lookup krbtgt/next_realm@our_realm */
+
+    krb5_free_principal(kcontext, norm_princ);
+    norm_princ = NULL;
+    kerr = krb5_build_principal_ext(kcontext, &norm_princ,
+                                    strlen(ipactx->realm),
+                                    ipactx->realm,
+                                    KRB5_TGS_NAME_SIZE,
+                                    KRB5_TGS_NAME,
+                                    strlen(trusted_realm),
+                                    trusted_realm, 0);
+    if (kerr != 0) {
+        goto done;
     }
 
+    kerr = dbget_princ(kcontext, ipactx, norm_princ, flags, entry);
+
 done:
     free(trusted_realm);
-    if ((kerr != 0) && (kentry != NULL)) {
-        ipadb_free_principal(kcontext, kentry);
-    }
-    ldap_msgfree(res);
+    krb5_free_principal(kcontext, norm_princ);
     krb5_free_unparsed_name(kcontext, principal);
+
     return kerr;
 }
 
+/* TODO: handle case where main object and krbprincipal data are not
+ * the same object but linked objects ?
+ * (by way of krbprincipalaux being in a separate object from krbprincipal).
+ * Currently we only support objcts with both objectclasses present at the
+ * same time. */
+
+krb5_error_code ipadb_get_principal(krb5_context kcontext,
+                                    krb5_const_principal search_for,
+                                    unsigned int flags,
+                                    krb5_db_entry **entry)
+{
+    struct ipadb_context *ipactx;
+    krb5_error_code kerr;
+
+    *entry = NULL;
+
+    ipactx = ipadb_get_context(kcontext);
+    if (!ipactx) {
+        return KRB5_KDB_DBNOTINITED;
+    }
+
+    if (!is_request_for_us(kcontext, ipactx->local_tgs, search_for)) {
+        return KRB5_KDB_NOENTRY;
```

Comment:
If we have any log facilities, it would be nice to log this.

Reply:
Added an info log entry.